### PR TITLE
[ACTION] Formstack new actions #1526

### PR DESCRIPTION
**Pull request by lcaresia**

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/formstack/formstack.app.mjs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+          handshake_key: this._accessToken(),
+        },
+      });
+    },
+    async removeWebhook(webhookId) {
+      return this._makeRequest(`webhooks/${webhookId}.json`, {
+        method: "delete",
+      });
+    },
+    async getForms({ $ }) {
```
Comment (jcortes) [inline components/formstack/formstack.app.mjs:58]: I would recommend creating the function `getForms` by just implementing what the description suggests in this case
```javascript
     async getForms(args = {}) {
       const {
         $,
         params,
       } = args;
       const options = {
         params,
       };
       return this._makeRequest("form", options, $);
     },
```
and then create a generic function to paginate over resources depending on the implemented function eg.
```javascript
    async paginateResources({
      resourceFn,
      resourceFnArgs = {},
      mapOnResponse = (response) => response.forms?.map((form) => form),
      maxPages = 10,
    }) {
      let page = resourceFnArgs.params?.page || 1;
      let allResources = [];

      do {
        const response =
          await resourceFn({
            ...resourceFnArgs,
            params: {
              page,
              per_page: 100,
              ...resourceFnArgs.params,
            },
          });

        const nextResources = mapOnResponse(response);

        page = nextResources.length === 100
          ? page + 1
          : 0;

        allResources = allResources.concat(nextResources);

      } while (page > 0 && page <= maxPages);

      return allResources;
    },
```
then you can paginate any resource like this
```javascript
    const forms = await this.paginateResources({
        resourceFn: this.getForms,
        resourceArgs: {
          params: {
            page: 1,
            per_page: 100,
          },
        },
        mapOnResponse: (response) => response.forms?.map((form) => form),
        maxPages: 10,
      });
```
Of course you can improve this strategy, but the idea is to be as generic as possible with code so we can reuse it later on.

File: components/formstack/actions/create-form/create-form.mjs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  props: {
+    formstack,
+    name: {
+      label: "Name",
+      description: "The form name",
+      type: "string",
+    },
+    fields: {
+      label: "Fields",
+      description: "Array of Field resources",
```
Comment (jcortes) [inline components/formstack/actions/create-form/create-form.mjs:18]: Can you expand a bit more the description here, so the user can get a better idea on how to input these values?, maybe we can change this to an array of strings? `type: "string[]"`.
Isn't this the same as the property in `create-form-submission` action component?

File: components/formstack/actions/create-form/create-form.mjs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        fields,
+        language,
+        db,
+        password,
+        submit_button_title: submitButtonTitle,
+      },
+      $,
+    });
+
+    $.export("summary", "Successfully created form");
```
Comment (jcortes) [inline components/formstack/actions/create-form/create-form.mjs:69]:
```suggestion
    $.export("summary", `Successfully created form with ID ${response.id}`);
```

File: components/formstack/actions/create-form-submission/create-form-submission.mjs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    },
+  },
+  async run({ $ }) {
+    const {
+      formId,
+      fields,
+      encryptionPassword,
+    } = this;
+
+    const parsedFields = JSON.parse(fields);
```
Comment (jcortes) [inline components/formstack/actions/create-form-submission/create-form-submission.mjs:36]: If user just input and empty string or a random character here seems like he is getting this error
```
Uncaught SyntaxError: Unexpected end of JSON input
```

File: components/formstack/sources/common/common.mjs [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import formstack from "../../formstack.app.mjs";
```
Comment (jcortes) [inline components/formstack/sources/common/common.mjs:1]: One convention I've used so far in terms of naming common files is that if you are in the `common` folder you should start with the `base.mjs` file name. So I would recommed rename `common/common.mjs` by `common/base.mjs`

**Review by jcortes — CHANGES_REQUESTED** [review]
Thanks @lcaresia it's looking great so far. I've just added a few request changes regarding reuse and minor questions/observations. Bu overall it looks good.

File: components/formstack/package.json [hunk 6]
```diff
@@ -10,9 +10,7 @@
     "homepage": "https://pipedream.com/apps/formstack",
     "author": "Pipedream <support@pipedream.com> (https://pipedream.com/)",
     "license": "MIT",
-    "dependencies": {
-        "axios": "^0.21.1"
-    },
+    "dependencies": {},
```
[the remark this answers is not in the document]
Comment (lcaresia) [reply components/formstack/package.json:13]: Why ? Now is used the `@pipedream/platform` axios.

File: components/formstack/actions/create-form-submission/create-form-submission.mjs (hunk 4 above)
Comment (lcaresia) [reply components/formstack/actions/create-form-submission/create-form-submission.mjs:36]: Have a default way (or an example) to validate this ?

File: components/formstack/actions/create-form/create-form.mjs (hunk 2 above)
Comment (lcaresia) [reply components/formstack/actions/create-form/create-form.mjs:18]: > Isn't this the same as the property in create-form-submission action component?

Isn't the same prop, one is just the field id and another is the field with values.

File: components/formstack/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (lcaresia) [reply components/formstack/package.json]: [Resolved here](https://github.com/PipedreamHQ/pipedream/pull/2669#discussion_r867071209)

File: components/formstack/actions/create-form-submission/create-form-submission.mjs (hunk 4 above)
Comment (jcortes) [reply components/formstack/actions/create-form-submission/create-form-submission.mjs]: The way I usually test this is by using the UI. Just by input some space in the input box.

**Review by michelle0927 — APPROVED** [review]
LGTM!

**Comment by DilanAthukorala** [comment]
@lcaresia Can you please add an example here of how to send fields, I'm getting following error when sending field types (name, email, address) 
image
image

**Comment by lcaresia** [comment]
@DilanAthukorala Example added! Call me if you have problems.

**Comment by DilanAthukorala** [comment]
@lcaresia Thanks. 

Verification Completed.

**Review by andrewjschuang — APPROVED** [review]

**Pull request merged**